tuna/yaml_parser.py

```diff
@@ -0,0 +1,107 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+"""! @brief YAML parser to support multiple tuning steps """
+
+import os
+import tempfile
+import yaml
+from tuna.libraries import Library
+from tuna.parse_args import TunaArgs
+from tuna.miopen.metadata import MIOPEN_TUNING_STEPS, MIOPEN_SINGLE_OP
+from tuna.example.metadata import EXAMPLE_TUNING_STEPS, EXAMPLE_SINGLE_OP
+
+
+def parse_yaml(filename, lib):
+  """Parses input yaml file and returns 1 or multiple yaml files(when library support
+     for multiple yaml files is provided).
+     Multiple yaml files are returned when more that 1 step per initial yaml file
+     are specified."""
+  yaml_dict = None
+  yaml_files = []
+  with open(os.path.expanduser(filename), encoding="utf8") as stream:
+    try:
+      yaml_dict = yaml.safe_load(stream)
+    except yaml.YAMLError as exc:
+      print(exc)
+
+  #if Library not specified here, no custom parsing function will be used
+  if lib == Library.MIOPEN or lib.value == 'miopen':
+    yaml_files = get_yaml_files(yaml_dict, MIOPEN_TUNING_STEPS,
+                                MIOPEN_SINGLE_OP)
+  elif lib == Library.EXAMPLE or lib.value == 'example':
+    yaml_files = get_yaml_files(yaml_dict, EXAMPLE_TUNING_STEPS,
+                                EXAMPLE_SINGLE_OP)
+  else:
+    #return current yaml file without custom parsing
+    return filename
+
+  return yaml_files
+
+
+def get_common_yaml(yaml_dict):
+  """Set TunaArgs common yaml fields"""
+  common_args = [enum.value for enum in TunaArgs]
+  common_yaml_part = {}
+
+  #extract common TunaArgs
+  for key in yaml_dict:
+    if key in common_args:
+      common_yaml_part[key] = yaml_dict[key]
+
+  return common_yaml_part
+
+
+def dump_yaml(new_yaml):
+  """Dump yaml file to /tmp"""
+  _, local_file = tempfile.mkstemp()
+  with open(local_file, 'w', encoding="utf8") as outfile:
+    yaml.dump(new_yaml, outfile, default_flow_style=False)
+
+  return local_file
+
+
+def get_yaml_files(yaml_dict, tuning_steps, single_op):
+  """Split initial YAML file to multiple files and return list of files"""
+  yaml_files = []
+  common_yaml_part = get_common_yaml(yaml_dict)
+
+  for key in yaml_dict:
+    #only look at enabled items
+    if key in tuning_steps and yaml_dict.get(key).get('enabled'):
+      #append common args
+      new_yaml = common_yaml_part.copy()
+      yaml_dict[key].pop('enabled', None)
+
+      #append nested dicts
+      for key2, value2 in yaml_dict.get(key).items():
+        new_yaml[key2] = value2
+
+      if key in single_op:
+        new_yaml[key] = True
+      yaml_files.append(dump_yaml(new_yaml))
```

Comment:
At these point they are yaml dicts?

Reply:
yes, the file gets split up in multiple dicts and then they get written to disk as separate yaml files, so basically splitting a yaml file with multiple tuning instructions into separate files with each 1 tuning instruction